onnxruntime/core/mlas/lib/halfgemm.h

```diff
@@ -0,0 +1,515 @@
+            K,
+            RangeStartN,
+            0);
+        ldb = MlasHalfGemmPackedBLeadingDim<KernelType>(N, K);
+    } else {
+        pb = reinterpret_cast<const _mlas_fp16_*>(Data->B) + RangeStartN;
+    }
+
+    const _mlas_fp16_* Bias = (nullptr == Data->Bias)
+        ? nullptr
+        : reinterpret_cast<const _mlas_fp16_*>(Data->Bias) + RangeStartN;
+    _mlas_fp16_* c = reinterpret_cast<_mlas_fp16_*>(Data->C)
+        + RangeStartM * ldc + RangeStartN;
+
+    size_t RowsRemaining = RangeCountM;
+    while (RowsRemaining > 0) {
+        MlasHalfGemmKernel<KernelType>(
+            RowsRemaining,
+            RangeCountN,
+            K,
+            c,
+            ldc,
+            Bias,
+            pa,
+            lda,
+            pb,
+            ldb,
+            true);
+
+        size_t RowsHandled = std::min(RowsRemaining, KernelType::KernelMaxM);
+
+        if (Data->OutputProcessor != nullptr) {
+            Data->OutputProcessor->Process(
+                Data->C,
+                RangeStartM + RangeCountM - RowsRemaining,
+                RangeStartN,
+                RowsHandled,
+                RangeCountN,
+                Data->ldc);
+        }
+
+        c += ldc * RowsHandled;
+        pa += lda * RowsHandled;
+        RowsRemaining -= RowsHandled;
+    }
+}
+
+
+template<typename KernelType>
+void
+MlasHalfGemmOperation(
+    const size_t N,
+    const size_t K,
+    const MLAS_HALF_GEMM_DATA_PARAMS* Data,
+    const size_t RangeStartM,
+    const size_t RangeCountM,
+    const size_t RangeStartN,
+    const size_t RangeCountN
+    )
+{
+    const size_t lda = Data->lda;
+    const size_t ldb = Data->ldb;
+    const size_t ldc = Data->ldc;
+
+    if (!Data->AIsfp32 && (ldb == 0 || (!KernelType::PackNeeded && !Data->BIsfp32))) {
+        // !Data->AIsfp32 => A is fp16, no packing on the left hand side
+        // ldb == 0 => B is already packed, no packing on the right hand side
+        // !KernelType::PackNeeded && !Data->BIsfp32  => B is fp16 and the kernel
+        //      does not require packing
+        //
+        // So no packing needed on either A or B, use a simpler driver instead
+
+        MlasHalfGemmNoPackOperation<KernelType>(
+            N,
+            K,
+            Data,
+            RangeStartM,
+            RangeCountM,
+            RangeStartN,
+            RangeCountN);
+        return;
+    }
+
+    const auto* Bias = reinterpret_cast<const _mlas_fp16_*>(Data->Bias);
+    _mlas_fp16_* C = reinterpret_cast<_mlas_fp16_*>(Data->C)
+        + RangeStartM * ldc + RangeStartN;
+
+    //
+    // Three dimensional tiling due to limited packing panel size
+    //
+    constexpr MLAS_HALF_GEMM_STRIDES Strides = KernelType::Strides;
+    constexpr size_t packASize = UpAlignSize(Strides.M * Strides.K * FP16_SIZE);
+    constexpr size_t packBSize = UpAlignSize(Strides.N * Strides.K * FP16_SIZE);
+    MlasThreadedBufAlloc(packASize + packBSize);
+
+    uint8_t* p = ThreadedBufHolder.get();
```

Comment:
>ThreadedBufHolder.get(); [](http://example.com/codeflow?start=17&length=24)

nit: make it return void*

Reply:
yeah I switched back and forth between uint8_t* and void* several times :-)  Almost always, the returned pointer needs to participate in pointer arithmetic for dividing the returned memory region into multiple buffers. Since the MlasThreadedBufAlloc function takes the buffer size in bytes, I thought it would be better the returned pointer type is in bytes to reflect this calculation.